### PR TITLE
Learning rate, betas, eps, weight_decay at model level

### PR DESCRIPTION
**Pull request by piercus**
Related to #165 where I'm trying to train a small encoder and a big adapter together (and putting different learning rate will help)

I need to confirm that is working on real-life training run, but if you have any feedback in the meanwhile it would be great.

Especially, this implementation covers learning rate, betas, eps, weight_decay but we could do only learning rate, it will cover 99% of the needs.

### CODE REVIEW

**Comment by limiteinductive** [comment]
Thanks for the PR; I'll have a look at it!

**Comment by limiteinductive** [comment]
That's a great addition, let's rebase and then we can merge!

**Review by limiteinductive — COMMENTED** [review]
Could you rebase?

File: src/refiners/training_utils/config.py [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-from typing import Any, Callable, Iterable, Literal, Type, TypeVar
+from typing import Any, Callable, Literal, Type, TypeVar
 
 import tomli
 from bitsandbytes.optim import AdamW8bit, Lion8bit  # type: ignore
 from prodigyopt import Prodigy  # type: ignore
 from pydantic import BaseModel, validator
-from torch.nn import Parameter
 from torch.optim import SGD, Adam, AdamW, Optimizer
+from torch.optim.optimizer import params_t as ParamsT
```
[the remark this answers is not in the document]
Comment (piercus) [reply src/refiners/training_utils/config.py:11]: sry I don't get it

`from torch.optim.optimizer import ParamsT` not working

[the remark this answers is not in the document]
Comment (limiteinductive) [reply src/refiners/training_utils/config.py:11]: or we could just write the Alias ourselves as `Params = Iterable[nn.Parameters] | Iterable[dict[str, Any]]

**Comment by piercus** [comment]
@limiteinductive please review

File: src/refiners/training_utils/config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (limiteinductive) [reply src/refiners/training_utils/config.py:11]:
```suggestion
from torch.optim.optimizer import ParamsT
```

**Review by limiteinductive — CHANGES_REQUESTED** [review]

File: src/refiners/training_utils/config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (deltheil) [reply src/refiners/training_utils/config.py:11]: OK, here is the thing re: this alias:

**PyTorch 2.1.0**

```python
params_t: TypeAlias = Union[Iterable[torch.Tensor], Iterable[Dict[str, Any]]]
```
See: [code](https://github.com/pytorch/pytorch/blob/v2.1.0/torch/optim/optimizer.py#L207)

**PyTorch 2.2.0**

```python
ParamsT: TypeAlias = Union[Iterable[torch.Tensor], Iterable[Dict[str, Any]]]
```
After this [PR](https://github.com/pytorch/pytorch/pull/111114)

[the remark this answers is not in the document]
Comment (deltheil) [reply src/refiners/training_utils/config.py]: @piercus at this point in time we would prefer not enforcing PyTorch 2.2+ (see https://github.com/finegrain-ai/refiners/commit/7eb8eb4c68fa54cc86f043ecce5d1feed192b5ef). So I am afraid you should undo https://github.com/finegrain-ai/refiners/commit/553cd181f933e77921301cccd99242b8c334ddb8

- Option 1: poor man/not-so-great like:

```python
try:
    from torch.optim.optimizer import params_t as ParamsT  # PyTorch 2.1. TODO: remove "soon"
except ImportError as e:
    from torch.optim.optimizer import ParamsT
``` 

- Option 2: would be just write (inline) the type alias ourselves as suggested by @limiteinductive 

Any other options that come to mind? cc @limiteinductive

[the remark this answers is not in the document]
Comment (limiteinductive) [reply src/refiners/training_utils/config.py]: not really, Option 2 doesn't work totally: this will create a type error when feeding it to the LRScheduler that expects the original type alias

[the remark this answers is not in the document]
Comment (piercus) [reply src/refiners/training_utils/config.py]: Option 1 done in [b6c3d65](https://github.com/finegrain-ai/refiners/pull/227/commits/b6c3d654781e84b45a4587429a44180d304cdb10)

**Review by deltheil — APPROVED** [review]
Nice!

File: src/refiners/training_utils/config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (deltheil) [reply src/refiners/training_utils/config.py]: Thanks Pierre!

**Review by limiteinductive — DISMISSED** [review]

**Comment by deltheil** [comment]
@piercus please have a look at the remaining points/suggestions! Then feel free to squash into a single commit (I can do it at merge time if you prefer) and we'd be good to go :)

**Comment by piercus** [comment]
@deltheil squash done

**Review by deltheil — APPROVED** [review]
